gpcontrib/arenadata_toolkit/arenadata_toolkit--1.4--1.5.sql

```diff
@@ -0,0 +1,121 @@
+/* gpcontrib/arenadata_toolkit/arenadata_toolkit--1.4--1.5.sql */
+
+/*
+ This is part of arenadata_toolkit API for ADB Bundle.
+ This function collects information in db_files_current and db_files_history tables.
+ */
+CREATE OR REPLACE FUNCTION arenadata_toolkit.adb_collect_table_stats()
+RETURNS VOID
+AS $$
+BEGIN
+	IF NOT EXISTS (
+		SELECT
+		FROM pg_partition pp
+		JOIN pg_class cl1 ON pp.parrelid = cl1.oid
+		JOIN pg_partition_rule pr ON pr.paroid = pp.oid
+		JOIN pg_class cl2 ON cl2.oid = pr.parchildrelid
+		JOIN pg_namespace n1 ON cl1.relnamespace = n1.oid
+		JOIN pg_namespace n2 ON cl2.relnamespace = n2.oid
+		WHERE
+			CASE
+				WHEN cl1.relname = 'db_files_history' AND
+					n1.nspname = 'arenadata_toolkit' AND
+					n2.nspname = 'arenadata_toolkit'
+				THEN now() BETWEEN 
+					CAST(substring(pg_get_expr(pr.parrangestart, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
+						AS TIMESTAMP WITHOUT TIME ZONE)
+					AND
+					CAST(substring(pg_get_expr(pr.parrangeend, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
+						AS TIMESTAMP WITHOUT TIME ZONE)

```

Comment:
> You can move filters n1.nspname = 'arenadata_toolkit' AND n2.nspname = 'arenadata_toolkit' from CASE to WHERE. It will do query plan more optimal

But in this case, if we have a table 'db_files_history' in other namespace, which is partitioned by bigint, we will get the error again, no?

Reply:
> > But in this case, if we have a table 'db_files_history' in other namespace, which is partitioned by bigint, we will get the error again, no?
> 
> I think no, but you may check this.

I've rechecked. If we have, for ex., 'public.db_files_history' table partitioned by bigint, it leads to an error.
So this option is not applicable.

> then consider using opclass

Ok, here is a working option, do you see any drawbacks?

```
postgres=# explain              SELECT
                FROM pg_partition pp
                JOIN pg_class cl1 ON pp.parrelid = cl1.oid
                JOIN pg_partition_rule pr ON pr.paroid = pp.oid
                JOIN pg_class cl2 ON cl2.oid = pr.parchildrelid
                JOIN pg_namespace n1 ON cl1.relnamespace = n1.oid
                JOIN pg_namespace n2 ON cl2.relnamespace = n2.oid
                WHERE
                        CASE
                                WHEN pp.parclass[0] IN (SELECT oid FROM pg_opclass WHERE opcname = 'timestamp_ops')
                                THEN now() BETWEEN
                                        CAST(substring(pg_get_expr(pr.parrangestart, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                                        AND
                                        CAST(substring(pg_get_expr(pr.parrangeend, pr.parchildrelid) FROM '#"%#"::%' FOR '#')
                                                AS TIMESTAMP WITHOUT TIME ZONE)
                                ELSE FALSE
                        END
                        AND
                        cl1.relname = 'db_files_history'
                        AND
                        n1.nspname = 'arenadata_toolkit'
                        AND
                        n2.nspname = 'arenadata_toolkit'
                        AND
                        pp.paristemplate = false
                        AND
                        pp.parkind = 'r'  --range
                        AND
                        pr.parrangestartincl = 't';
                                                                                                                                                                                                                                 QUERY PLA
N                                                                                                                                                                                                                                  
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Hash Join  (cost=20.42..33.50 rows=4 width=0)
   Hash Cond: (cl2.relnamespace = n2.oid)
   ->  Hash Join  (cost=19.32..32.36 rows=4 width=4)
         Hash Cond: (cl2.oid = pr.parchildrelid)
         ->  Seq Scan on pg_class cl2  (cost=0.00..11.37 rows=437 width=8)
         ->  Hash  (cost=19.28..19.28 rows=2 width=4)
               ->  Hash Join  (cost=6.66..19.28 rows=4 width=4)
                     Hash Cond: (cl1.relnamespace = n1.oid)
                     ->  Hash Join  (cost=5.56..18.13 rows=4 width=8)
                           Hash Cond: (pp.oid = pr.paroid)
                           Join Filter: CASE WHEN (hashed SubPlan 1) THEN (('2024-04-11 11:43:22.573817+10'::timestamp with time zone >= ("substring"(pg_get_expr(pr.parrangestart, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without time zone) AND ('2024-04-11 11:43:22.573817+10'::timestamp with time zone <= ("substring"(pg_get_expr(pr.parrangeend, pr.parchildrelid), '#"%#"::%'::text, '#'::text))::timestamp without time zone)) ELSE false END
                           ->  Hash Join  (cost=1.01..13.51 rows=4 width=40)
                                 Hash Cond: (cl1.oid = pp.parrelid)
                                 ->  Seq Scan on pg_class cl1  (cost=0.00..12.46 rows=1 width=8)
                                       Filter: (relname = 'db_files_history'::name)
                                 ->  Hash  (cost=1.00..1.00 rows=1 width=40)
                                       ->  Seq Scan on pg_partition pp  (cost=0.00..1.00 rows=1 width=40)
                                             Filter: ((NOT paristemplate) AND (parkind = 'r'::"char"))
                           ->  Hash  (cost=1.00..1.00 rows=1 width=72)
                                 ->  Seq Scan on pg_partition_rule pr  (cost=0.00..1.00 rows=1 width=72)
                                       Filter: parrangestartincl
                           SubPlan 1
                             ->  Materialize  (cost=0.00..3.54 rows=4 width=4)
                                   ->  Seq Scan on pg_opclass  (cost=0.00..3.53 rows=4 width=4)
                                         Filter: (opcname = 'timestamp_ops'::name)
                     ->  Hash  (cost=1.09..1.09 rows=1 width=4)
                           ->  Seq Scan on pg_namespace n1  (cost=0.00..1.09 rows=1 width=4)
                                 Filter: (nspname = 'arenadata_toolkit'::name)
   ->  Hash  (cost=1.09..1.09 rows=1 width=4)
         ->  Seq Scan on pg_namespace n2  (cost=0.00..1.09 rows=1 width=4)
               Filter: (nspname = 'arenadata_toolkit'::name)
 Optimizer: Postgres query optimizer
(32 rows)
```

Reply:
> I've rechecked. If we have, for ex., 'public.db_files_history' table partitioned by bigint, it leads to an error.

may be add it to test (or rename part_table_2 to db_files_history)?

Reply:
Yes, it would be useful. Added.